### PR TITLE
Audit IORefs

### PR DESCRIPTION
**Pull request by robdockins**
Audit all uses of IORefs in What4 and replace with with atomic primitives where necessary, or document when they require external synchronization.

### CODE REVIEW

File: what4/src/What4/Expr/Builder.hs [hunk 1]
```diff
@@ -8,6 +8,19 @@ Maintainer  : jhendrix@galois.com
 This module defines the canonical implementation of the solver interface
 from "What4.Interface". Type @'ExprBuilder' t st@ is
 an instance of the classes 'IsExprBuilder' and 'IsSymExprBuilder'.
+
+Notes regarding concurrency: The expression builder datatype contains
+a number of mutable storage locations.  These are designed so they
+may reasonably be used in a multithreaded context.  In particular,
+nonce values are generated atomically, and other IORefs used in this
+module are modified or written atomically, so modifications should
+propigate in the expected sequentually-consistent ways.  Of course,
```
Comment (brianhuffman) [inline what4/src/What4/Expr/Builder.hs:17]: "propagate", "sequentially"

File: what4/src/What4/Expr/Builder.hs [hunk 2]
```diff
@@ -8,6 +8,19 @@ Maintainer  : jhendrix@galois.com
...
+threads may still clobber state others have set (e.g., the current 
+program location) so the potentialy for truly multithreaded us is
```
Comment (brianhuffman) [inline what4/src/What4/Expr/Builder.hs:19]: "potential for truly multithreaded use"

File: what4/src/What4/Config.hs [hunk 3]
```diff
@@ -64,6 +64,16 @@
...
 --       (i.e., to undo extendConfig)
+--
+--
+-- Note regarding concurrency: the configuration data structures in this
+-- module are implemented using MVars, and may safely be used in a multithreaded
+-- way; configuration changes made in one thread will be visible to others
+-- in a properly synchronized way.  Of course, if one desires to isolate
+-- configuration changes in different threads from each other, separate
+-- configuration objects are required. As noted in the documentation for
+-- @set_opt_onset@, the validation procedures for options should not
```
Comment (brianhuffman) [inline what4/src/What4/Config.hs:75]: Maybe write `'set_opt_onset'` so that the haddock will have a hyperlink.

File: what4/src/What4/Config.hs [hunk 4]
```diff
@@ -657,17 +669,17 @@ insertOption (ConfigDesc (ConfigOption _tp (p:|ps)) sty h) m = adjustConfigMap p
 ------------------------------------------------------------------------
 -- Config
 
--- | The main configuration datatype.  It consists of an IORef
+-- | The main configuration datatype.  It consists of an MVar
 --   continaing the actual configuration data.
```
Comment (brianhuffman) [inline what4/src/What4/Config.hs:673]: "containing"

**Review by brianhuffman — APPROVED** [review]
Aside from some haddock nitpicking, it all looks very reasonable.

**Comment by robdockins** [comment]
Build failure appears to be due to this: https://github.com/Gabriel439/Haskell-MMorph-Library/issues/54

**Pull request merged**